### PR TITLE
KAFKA-17482: Make share partition initialization async (KIP-932)

### PR DESCRIPTION
**Pull request by apoorvmittal10**
The PR introduces states for share partition which are used to make share partition initilzation async.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by apoorvmittal10** [comment]
@AndrewJSchofield @adixitconfluent @omkreddy Build passed with unrelated tests failure. Can you please take a look.

File: core/src/main/java/kafka/server/share/SharePartitionManager.java [hunk 1]
```diff
@@ -626,6 +613,85 @@ void maybeProcessFetchQueue() {
...
+        if (throwable instanceof LeaderNotAvailableException) {
+            log.debug("The share partition with key {} is not initialized yet", sharePartitionKey);
+            // Do not process the fetch request for this partition as the leader is not initialized yet.
+            // The fetch request will be retried in the next poll.
+            // TODO: Add the request to delayed fetch purgatory.
+            return;
+        }
+
+        if (throwable instanceof NotLeaderOrFollowerException || throwable instanceof FencedStateEpochException) {
+            log.info("The share partition with key {} is fenced with message: {}", sharePartitionKey, throwable.getMessage());
```
Comment (AndrewJSchofield) [inline core/src/main/java/kafka/server/share/SharePartitionManager.java]: nit: I'd remove the "with message". Still makes sense that you're adding the exception message after the colon.

**Review by AndrewJSchofield — APPROVED** [review]
Thanks for the PR. I note the comment about the future use of the purgatory to properly handle initialization which exceeds the duration of the desired wait time. That's important and coming in a later PR. Subject to that proviso, approved.

File: core/src/main/java/kafka/server/share/SharePartitionManager.java (hunk 1 above)
Comment (apoorvmittal10) [reply core/src/main/java/kafka/server/share/SharePartitionManager.java]: Thanks, done.

**Comment by apoorvmittal10** [comment]
@AndrewJSchofield @omkreddy Build passed with unrelated test failures.

**Review by omkreddy — APPROVED** [review]
@apoorvmittal10 Thanks for the PR. LGTM

**Pull request merged**